File: services/payment-service/internal/infrastructure/messaging/rabbitmq/connection.go
package rabbitmq

import (
	"log"
	"os"

	amqp "github.com/rabbitmq/amqp091-go"
)

type Connection struct {
	conn *amqp.Connection
	channel *amqp.Channel
	url string
}

func NewConnection() *Connection {
	url, exists := os.LookupEnv("RABBITMQ_URL")

	if !exists {
		log.Fatal("A constante da URL do RabbitMQ URL não foi encontrada!")
	}

	c := &Connection{url: url}
	c.connect()
	return c

}

func (c *Connection) connect() {
	conn, err := amqp.Dial(c.url)
	if err != nil {
		log.Fatal("Não foi possível estabelecer a conexão com o RabbitMQ")
	}

	channel, err := conn.Channel()
	if err != nil {
		conn.Close()
		log.Fatal("Não foi possível abrir uma canal no RabbitMQ")
	}

	c.conn = conn
	c.channel = channel

	err = setup(channel)
	if err != nil {
		log.Fatal("Erro ao configurar topologia do RabbitMQ")
	}
}

func (c *Connection) Channel() *amqp.Channel {
	if c.channel == nil || c.channel.IsClosed() {
		channel, err := c.conn.Channel()
		if err != nil {
			log.Println("Erro ao abrir canal no RabbitMQ!")
			c.connect()
			return c.channel
		}
		log.Println("Canal aberto no RabbitMQ!")
		c.channel = channel

	}

	log.Println("Já existe um canal aberto no RabbitMQ!")
	return c.channel
}

func (c *Connection) Close() {
	if c.channel != nil && !c.channel.IsClosed() {
		c.channel.Close()
	}
	if c.conn != nil && !c.conn.IsClosed() {
		c.conn.Close()
	}
	log.Println("Conexão com o RabbitMQ encerrada!")
}

File: services/estoque/app/infrastructure/messaging/rabbitmq/connection.py
import os
import logging
import aio_pika
from aio_pika import Connection, Channel

from .setup import setup

logger = logging.getLogger(__name__)


class RabbitMQConnection:
    def __init__(self):
        self._url = os.environ.get("RABBITMQ_URL")
        if not self._url:
            raise EnvironmentError("A constante da URL do RabbitMQ não foi encontrada!")

        self._connection: Connection | None = None
        self._channel: Channel | None = None

    async def connect(self):
        self._connection = await aio_pika.connect_robust(self._url)
        self._channel = await self._connection.channel()
        await setup(self._channel)
        logger.info("Conexão com o RabbitMQ estabelecida!")

    async def channel(self) -> Channel:
        if self._channel is None or self._channel.is_closed:
            if self._connection is None or self._connection.is_closed:
                await self.connect()
            else:
                self._channel = await self._connection.channel()
                logger.info("Novo canal aberto no RabbitMQ!")
        else:
            logger.info("Já existe um canal aberto no RabbitMQ!")

        return self._channel

    async def close(self):
        if self._channel and not self._channel.is_closed:
            await self._channel.close()
        if self._connection and not self._connection.is_closed:
            await self._connection.close()
        logger.info("Conexão com o RabbitMQ encerrada!")
